### ISSUE TITLE
fix for warning: documentation for unknown define

### ISSUE THREAD

**Issue by combsie** [doxygen/doxygen#9195]
I am having this pop up as an error with running doxygen:

> warning: documentation for unknown define

This is my *.dox file for preprocessing:
```
#---------------------------------------------------------------------------
# Configuration options related to the preprocessor
#---------------------------------------------------------------------------

# If the ENABLE_PREPROCESSING tag is set to YES, doxygen will evaluate all
# C-preprocessor directives found in the sources and include files.
# The default value is: YES.

ENABLE_PREPROCESSING   = YES

# If the MACRO_EXPANSION tag is set to YES, doxygen will expand all macro names
# in the source code. If set to NO, only conditional compilation will be
# performed. Macro expansion can be done in a controlled way by setting
# EXPAND_ONLY_PREDEF to YES.
# The default value is: NO.
# This tag requires that the tag ENABLE_PREPROCESSING is set to YES.

MACRO_EXPANSION        = YES

# If the EXPAND_ONLY_PREDEF and MACRO_EXPANSION tags are both set to YES then
# the macro expansion is limited to the macros specified with the PREDEFINED and
# EXPAND_AS_DEFINED tags.
# The default value is: NO.
# This tag requires that the tag ENABLE_PREPROCESSING is set to YES.

EXPAND_ONLY_PREDEF     = YES
```
And this is the code snippet:
```
/**
 *  @def MacroA(mType, mStartLane, mEndLane, mResetId) {mType, 0, 0, mStartLane, mEndLane, mResetId, 0, 0, 0}
 *  @brief Macro for statically initializing various structures.
 *  @param [in] mType Specifies the blah.
 *  @param [in] mStartLane Identifies the socket-relative Start Lane (lowest numbered lane unless reversed)
 *  @param [in] mEndLane Identifies the socket-relative End Lane
 *  @param [in] mResetId A value that identifies a specific GPIO that provides PERST#.
 *  @details This is a simplified engine descriptor that has removed unused fields (see MacroB for past implementation)
 */
#define  MacroA(mType, mStartLane, mEndLane, mResetId) {mType, 0, 0, mStartLane, mEndLane, mResetId, 0, 0, 0}
```
Any help would be greatly appreciated

**Comment by albert-github** [comment]
With the information provided and the current doxygen version (1.9.3 (c0b9eafbfb53286ce31e75e2b6c976ee4d345473)) I cannot reproduce the problem.

- Can you please attach a, small, self contained example (source+configuration file in a tar or zip) that allows us to reproduce the problem? Please don't add external links as they might not be persistent.
- Please also specify the **full** doxygen version used (`doxygen -v`).

**Comment by combsie** [comment]
Here is the doxygen -v

C:\>doxygen -v
1.9.3 (c0b9eafbfb53286ce31e75e2b6c976ee4d345473)

Getting an example can be pretty tedious.  Is there any documentation that can explain why "warning: documentation for unknown define" and/or "warning: Documentation for undefined enum" can occur?

**Comment by ferdymercury** [comment]
> Here is the doxygen -v
> 
> C:>doxygen -v 1.9.3 (c0b9eafbfb53286ce31e75e2b6c976ee4d345473)
> 
> Getting an example can be pretty tedious. Is there any documentation that can explain why "warning: documentation for unknown define" and/or "warning: Documentation for undefined enum" can occur?

I guess it means that Doxygen found a documentation block that is supposed to be linked to a "enum X", but then it does not find the place in the C++ code that actually declares that code.

Please try using instead:
`@def MacroA(mType, mStartLane, mEndLane, mResetId)`
without the curly brackets.

Also: maybe run doxygen with the preprocessor flag active, to see the output.

**Comment by combsie** [comment]
How do I run doxygen with the preprocessor flag active?

**Comment by ferdymercury** [comment]
doxygen -d preprocessor

**Comment by combsie** [comment]
ok in the doxy.log
source.h:1119: warning: documentation for unknown define MacroA found.

in the source.html file:
<div class="line"><a id="l01128" name="l01128"></a><span class="lineno"> 1128</span><span class="preprocessor">#define  MacroA(mType, mStartLane, mEndLane, mResetId) {mType, 0, 0, mStartLane, mEndLane, mResetId, 0, 0, 0}</span></div>

In the source.xml file:
<div class="line"><a id="l01128" name="l01128"></a><span class="lineno"> 1128</span><span class="preprocessor">#define  MacroA(mType, mStartLane, mEndLane, mResetId) {mType, 0, 0, mStartLane, mEndLane, mResetId, 0, 0, 0}</span></div>

And we have an rst file cause we are using sphynz and breathe to put together a pdf:
.. doxygendefine:: MacroA

**Comment by albert-github** [comment]
See https://github.com/doxygen/doxygen/issues/9195#issuecomment-1064149272

**Comment by combsie** [comment]
So I've been doing some more searching and my project uses sphix and breathe to make a PDF. In one of the rst files we use:

.. doxygendefine:: MacroA
    :project: ProjectX

In the PDF this results in:

Warning: doxygendefine: Cannot find define “MacroA” in doxygen xml output for project
“ProjectX” from directory: build/DoxyScans/xml/

I saw this example (https://github.com/michaeljones/breathe/blob/master/documentation/source/define.rst) but still don't understand the what I'm doing wrong. Hopefully this moves us further down the path of success

**Comment by albert-github** [comment]
Still missing an example as requested in https://github.com/doxygen/doxygen/issues/9195#issuecomment-1063850586 without it it is not possible to help. The most likely cause is that there is a problem on the sphinx / breathe side definitions resulting in a wrong interpretation of the doxygen xml results (so probably the issue is not an doxygen issue).

**Comment by combsie** [comment]
Ok here is an example I made. You'll need breathe and latex installed as well as doxygen. Hopefully you can see the error in the PDF produced
[ProjectX_GitHub.zip](https://github.com/doxygen/doxygen/files/8499294/ProjectX_GitHub.zip)
.

**Comment by albert-github** [comment]
As far as I know breathe/sphinx use the doxygen xml output for further processing, but neither in the `ProjectX.dox` nor in the `genpdf.bat` I see any evidence that the `GENERATE_XML` us set to `YES`

**Comment by combsie** [comment]
Also to get the output from a cmd prompt run 'genpdf.bat ProjectX' in the ProjectX_GitHub\ProjectX\ProjectDocs dir. I did change the GENERATE_XML to YES and still got the same results. Are there any other settings that I need to adjust?

**Comment by albert-github** [comment]
No idea, I don't know breathe / sphinx, they should specify what is necessary. Maybe you need to set e.g. `INPUT` or `RECURSIVE` or ...

**Comment by okmechak** [comment]
I have just added new empty line after #def.. at this resolved my problem ;)
![image](https://user-images.githubusercontent.com/5883082/165802087-b323a760-a52f-4000-bb8d-7d26d75fe916.png)

**Comment by albert-github** [comment]
I had a look at the problem as @okmechak indicated and had to make the assumption that the `#define VERSION_PATH 0` is the last line in the file and that it is not followed by a line ending i.e. `\n` (or `\r\n` or `\r`). It is known that some editors have this habit.
With the current doxygen version 1.9.3 and with the current master version (Doxyfile 1.9.4 (fe58686503d386afb5269e7a2799c0457c743092)) I can reproduce this problem.

My example: [example.tar.gz](https://github.com/doxygen/doxygen/files/8589577/example.tar.gz)

**Comment by albert-github** [comment]
I've just pushed a proposed patch, pull request #9300.

My new example (also a DOS end of line file): [example1.tar.gz](https://github.com/doxygen/doxygen/files/8592571/example1.tar.gz)

**Comment by albert-github** [comment]
Code has been integrated in master on GitHub (please don't close the issue as this will be done at the moment of an official release).

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.9.4.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
